guvnor-project/guvnor-project-builder/src/main/java/org/guvnor/common/services/builder/ResourceChangeObserver.java

```diff
@@ -147,12 +130,40 @@ private void processResourceChange( final SessionInfo sessionInfo,
                                   + ", changeType: " + changeType );
         }
 
-        if ( !notifiedProjects.containsKey( project.getRootPath().toURI() ) && isObservableResource( path ) ) {
+        if ( isObservableResource( path ) ) {
             invalidateDMOProjectCacheEvent.fire( new InvalidateDMOProjectCacheEvent( sessionInfo,
                                                                                      project,
                                                                                      path ) );
-            notifiedProjects.put( project.getRootPath().toURI(),
-                                  Boolean.TRUE );
+        }
+    }
+
+    private void processBatchResourceChanges( final SessionInfo sessionInfo,
+                                              Map<Path, Collection<ResourceChange>> resourceChanges ) {
+
+        Map<Project, Path> pendingNotifications = new HashMap<Project, Path>( );
+        Project project;
+
+        for ( final Map.Entry<Path, Collection<ResourceChange>> pathCollectionEntry : resourceChanges.entrySet() ) {
+
+            //Only process Project resources
+            project = projectService.resolveProject( pathCollectionEntry.getKey() );
+            if ( project == null ) {
+                continue;
+            }
+
+            if ( !pendingNotifications.containsKey( project ) && isObservableResource( pathCollectionEntry.getKey() ) ) {
```

Comment:
if first file for a given project is TheFile.java, the this file is putted on the hash, but then if pom.xml is processed then we want to remember it instead of the initial TheFile.java.